aten/src/ATen/native/cuda/MultinomialKernel.cu

```diff
@@ -2,13 +2,506 @@
 #include <ATen/NativeFunctions.h>
 #include <ATen/LegacyTHFunctionsCUDA.h>
 #include <ATen/native/UnaryOps.h>
+#include <ATen/cuda/CUDAContext.h>
+#include <ATen/cuda/CUDAApplyUtils.cuh>
+#include <ATen/native/cuda/LaunchUtils.h>
+#include <ATen/AccumulateType.h>
+
+#include <THC/THCReduceApplyUtils.cuh>
+#include <THC/THCTensorMathReduce.cuh>
+#include <THC/THCNumerics.cuh>
+
+#include <curand.h>
+#include <curand_kernel.h>
+#include <curand_philox4x32_x.h>
 
 namespace at { namespace native {
 
-void multinomial_kernel_impl(Tensor& result, const Tensor& self, const int64_t num_samples, const bool replacement, Generator* generator) {
-  legacy::cuda::_th_multinomial_out(result, self, num_samples, replacement, generator);
+namespace {
+
+#define MAX_NUM_BLOCKS 200
+
+// Normalizes the L1 norm of every row to 1; used by multinomial
+template <typename scalar_t>
+#ifdef __HIP_PLATFORM_HCC__
+C10_LAUNCH_BOUNDS_1(1024)
+#endif
+__global__ void renormRowsL1(scalar_t* dist, long rows, long cols) {
+  extern __shared__  unsigned char my_smem[];
+  scalar_t *smem = reinterpret_cast<scalar_t *>(my_smem);
+  scalar_t zero = static_cast<scalar_t>(0);
+  scalar_t val;
+  for (int64_t row = blockIdx.x; row < rows; row += gridDim.x) {
+    scalar_t sum = static_cast<scalar_t>(0);
+    for (int64_t col = threadIdx.x; col < cols; col += blockDim.x) {
+      val = dist[row * cols + col];
+      assert(!THCNumerics<scalar_t>::lt(val, zero)); // ! < 0 for NaN handling
```

Comment:
We're still using "THCNumerics" so this isn't 100% ported over. Is this intentional?

Reply:
`THCNumerics` removal seem to become completely separate task as it is used across aten in multiple occasions.

Reply:
So recommendation is to avoid using it where is it easy, and leave it as is where it is harder, to be removed later on as separate PR.

Reply:
I agree it's a separate task; I was curious what the plan for that was